### ISSUE TITLE
Incorrect SELinux labels on Qubes-provided kernel module directory

### ISSUE THREAD

**Issue by DemiMarie** [QubesOS/qubes-issues#4278]
### Qubes OS version:

R4.0

### Affected component(s): 

dom0 block devi

---

### Steps to reproduce the behavior:

Enable SELinux in a TemplateVM (only tested with Fedora 28 so far).

### Expected behavior:

Kernel modules load normally

### Actual behavior:

Kernel modules fail to load

### General notes:

Incorrect SELinux labelling in the filesystem mounted from `/var/lib/qubes/vm-kernels/*/modules.img` is the culprit.

---

### Related issues:

#4279, #4239

**Comment by marmarek** [comment]
The modules.img is built [here](https://github.com/QubesOS/qubes-linux-kernel/blob/master/kernel.spec#L460-L473), at `kernel-qubes-vm` package installation time in dom0. Feel free to provide a patch adding proper labeling. Note that selinux may not be available in dom0 at that time (and even if it is, policy version may be totally different than in VM).

On the other hand, the problem shouldn't apply to kernel installed in VM (`virt_mode=hvm`, `kernel=`).

**Comment by RWGuy** [comment]
@DemiMarie can you provide the list of changes you had to implement for selinux to work?

**Comment by DemiMarie** [comment]
@RWGuy

> @DemiMarie can you provide the list of changes you had to implement for selinux to work?

* I wrote a custom SELinux policy module, which allows QubesOS’s daemons the permissions they need.  It also runs `qubes-gui-runuser` as `xdm_t` and `qrexec-agent` as `local_login_t`, so `pam_selinux.so` works correctly.  Finally, it allows various programs access to `unlabeled_t`.
* I turned on polyinstantiation of `/tmp` and `/var/tmp`.  This is not strictly necessary, but it did force me to make `pam_selinux.so` work properly.  It is also necessary for maximum protection by SELinux, as QubesOS does not do anything to protect against `/tmp`-based symlink attacks.
* I had to modify the PAM configs to ensure that `pam_namespace.so` and `pam_selinux.so` are used by `su`, `qrexec-agent`, and `qubes-gui-runuser`.
* I had to add a drop-in file for `gnome-terminal-server.service` to pass `DISPLAY=:0` in the environment.  Otherwise, it did not start reliably.
